File: start_ngrok.py
import os
import re
import subprocess
import sys
import time
import urllib.request
import json

# Directory containing the HTML and JS files
ASSETS_DIR = os.path.join(os.path.dirname(__file__), 'assets', 'blockly')
ROOT_DIR = os.path.dirname(__file__)

def update_files(new_url):
    print(f"\n=======================================================")
    print(f"[+] Found new Ngrok URL: {new_url}")
    print(f"[+] Injecting new URL into frontend files...")
    
    count = 0
    # Regex to match CONVERT_SERVER assignment
    url_pattern = re.compile(r"const\s+CONVERT_SERVER\s*=\s*['\"].*?['\"];")
    replacement = f"const CONVERT_SERVER = '{new_url}';"
    
    files_to_check = []
    
    # Check assets/blockly/train.html
    train_html = os.path.join(ASSETS_DIR, 'train.html')
    if os.path.exists(train_html):
        files_to_check.append(train_html)
                
    # Check TrainDeployScreen.js
    train_js = os.path.join(ROOT_DIR, 'TrainDeployScreen.js')
    if os.path.exists(train_js):
        files_to_check.append(train_js)

    for filepath in files_to_check:
        try:
            with open(filepath, 'r', encoding='utf-8') as f:
                content = f.read()
            
            if url_pattern.search(content):
                new_content = url_pattern.sub(replacement, content)
                with open(filepath, 'w', encoding='utf-8') as f:
                    f.write(new_content)
                print(f"    -> Updated {os.path.basename(filepath)}")
                count += 1
        except Exception as e:
            print(f"[-] Error reading/writing {os.path.basename(filepath)}: {e}")
                
    print(f"[+] Successfully updated {count} files!")
    print(f"[*] You can now start 'npx expo start --tunnel'")
    print(f"=======================================================\n")

def get_ngrok_url():
    try:
        req = urllib.request.Request("http://127.0.0.1:4040/api/tunnels")
        with urllib.request.urlopen(req) as response:
            data = json.loads(response.read().decode())
            for tunnel in data.get('tunnels', []):
                if tunnel.get('public_url', '').startswith('https://'):
                    return tunnel['public_url']
    except Exception:
        return None
    return None

def main():
    print("[*] Starting Ngrok Tunnel on port 5001...")
    
    # Start ngrok process silently
    try:
        process = subprocess.Popen(
            ['ngrok', 'http', '5001'],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL
        )
    except FileNotFoundError:
        print("[-] Error: 'ngrok' command not found. Please install ngrok or add it to your PATH.")
        return

    print("[*] Waiting for Ngrok to initialize...")
    new_url = None
    # Poll for up to 10 seconds
    for _ in range(10):
        time.sleep(1)
        new_url = get_ngrok_url()
        if new_url:
            break

    if new_url:
        update_files(new_url)
        print("[*] Ngrok is running in the background. Press Ctrl+C to stop.")
        try:
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            print("\n[*] Stopping Ngrok tunnel...")
            process.terminate()
    else:
        print("[-] Failed to get Ngrok URL. Is another instance of ngrok already running?")
        process.terminate()

if __name__ == '__main__':
    main()

File: start_ngrok.js
const { spawn } = require('child_process');
const http = require('http');
const fs = require('fs');
const path = require('path');

const ROOT_DIR = __dirname;
const ASSETS_DIR = path.join(ROOT_DIR, 'assets', 'blockly');

function updateFiles(newUrl) {
    console.log(`\n=======================================================`);
    console.log(`[+] Found new Ngrok URL: ${newUrl}`);
    console.log(`[+] Injecting new URL into frontend files...`);

    let count = 0;
    const urlPattern = /const\s+CONVERT_SERVER\s*=\s*['"].*?['"];/g;
    const replacement = `const CONVERT_SERVER = '${newUrl}';`;

    const filesToCheck = [];

    const trainHtml = path.join(ASSETS_DIR, 'train.html');
    if (fs.existsSync(trainHtml)) filesToCheck.push(trainHtml);

    const trainJs = path.join(ROOT_DIR, 'TrainDeployScreen.js');
    if (fs.existsSync(trainJs)) filesToCheck.push(trainJs);

    for (const filepath of filesToCheck) {
        try {
            let content = fs.readFileSync(filepath, 'utf8');
            if (urlPattern.test(content)) {
                content = content.replace(urlPattern, replacement);
                fs.writeFileSync(filepath, content, 'utf8');
                console.log(`    -> Updated ${path.basename(filepath)}`);
                count++;
            }
        } catch (err) {
            console.log(`[-] Error reading/writing ${path.basename(filepath)}: ${err.message}`);
        }
    }

    console.log(`[+] Successfully updated ${count} files!`);
    console.log(`[*] You can now start 'npx expo start --tunnel'`);
    console.log(`=======================================================\n`);
}

function getNgrokUrl() {
    return new Promise((resolve) => {
        const req = http.get('http://127.0.0.1:4040/api/tunnels', (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                try {
                    const parsed = JSON.parse(data);
                    for (const tunnel of (parsed.tunnels || [])) {
                        if (tunnel.public_url && tunnel.public_url.startsWith('https://')) {
                            return resolve(tunnel.public_url);
                        }
                    }
                    resolve(null);
                } catch (e) {
                    resolve(null);
                }
            });
        });
        req.on('error', () => resolve(null));
        req.end();
    });
}

async function main() {
    console.log(`[*] Starting Ngrok Tunnel on port 5001...`);
    
    // Start ngrok process and stream output so we can see if it is downloading
    const ngrokProcess = spawn('npx --yes ngrok http 5001', {
        stdio: 'inherit',
        shell: true
    });

    console.log(`[*] Waiting for Ngrok to initialize...`);
    
    let newUrl = null;
    // Wait up to 60 seconds in case npx needs to download ngrok
    for (let i = 0; i < 60; i++) {
        await new Promise(r => setTimeout(r, 1000));
        newUrl = await getNgrokUrl();
        if (newUrl) break;
    }

    if (newUrl) {
        updateFiles(newUrl);
        console.log(`[*] Ngrok is running in the background. Press Ctrl+C to stop it.`);
        
        // Keep the process alive so ngrok doesn't die
        setInterval(() => {}, 1000000);
        
        process.on('SIGINT', () => {
            console.log(`\n[*] Stopping script...`);
            ngrokProcess.kill();
            process.exit(0);
        });
    } else {
        console.log(`[-] Failed to get Ngrok URL. Make sure npx completed successfully.`);
        ngrokProcess.kill();
        process.exit(1);
    }
}

main();
